src/mame/taito/taito_f3_v.cpp

```diff
@@ -1030,44 +1073,83 @@ bool taito_f3_state::mix_line(Mix &gfx, mix_pix *z, pri_mode *pri, const f3_line
 	constexpr int DEBUG_Y = 180 + V_START;
 	if (TAITOF3_VIDEO_DEBUG && line.y == DEBUG_Y) {
 		logerror("[%X] %s%d: %d,%d (%d)\n   {pal: %x/%x, blend: %x/%x, prio: %x/%x}\n",
-				 gfx.prio, gfx.debug_name(), gfx.debug_index,
+				 gfx.prio, gfx.debug_name(), gfx.index,
 				 gfx.blend_b(), gfx.blend_a(), gfx.blend_select(flags, 82),
-				 z[DEBUG_X].src_pal, z[DEBUG_X].dst_pal,
-				 z[DEBUG_X].src_blend, z[DEBUG_X].dst_blend,
-				 pri[DEBUG_X].src_prio, pri[DEBUG_X].dst_prio);
+				 z.src_pal[DEBUG_X], z.dst_pal[DEBUG_X],
+				 z.src_blend[DEBUG_X], z.dst_blend[DEBUG_X],
+				 pri.src_prio[DEBUG_X], pri.dst_prio[DEBUG_X]);
 	}
 
 	return false; // TODO: determine when we can stop drawing?
 }
 
-void taito_f3_state::render_line(pen_t *dst, const mix_pix (&z)[H_TOTAL])
+void taito_f3_state::render_line(pen_t *RESTRICT dst, const mix_pix &z)
 {
-	const pen_t *clut = &m_palette->pen(0);
-	for (int x = H_START; x < H_START + H_VIS; x++) {
-		const mix_pix mix = z[x];
-		rgb_t s_rgb = clut[mix.src_pal];
-		rgb_t d_rgb = clut[mix.dst_pal];
-		dst[x] = (s_rgb.scale8(mix.src_blend) + d_rgb.scale8(mix.dst_blend)).set_a(255);
+	const pen_t *clut = m_palette->pens();
+	for (unsigned int x = H_START; x < H_START + H_VIS; x++) {
+		rgb_t s_rgb = clut[z.src_pal[x]];
+		rgb_t d_rgb = clut[z.dst_pal[x]];
+
+		// source_color * src_blend + dest_color * dst_blend
+		u16 r1 = s_rgb.r();
+		u16 g1 = s_rgb.g();
+		u16 b1 = s_rgb.b();
+		u16 r2 = d_rgb.r();
+		u16 g2 = d_rgb.g();
+		u16 b2 = d_rgb.b();
+		r1 *= z.src_blend[x]; // these blend contributions have fixed3 precision
+		g1 *= z.src_blend[x]; // i.e. 0 (b0'000) to 8 (b1'000) represents 0.0 to 1.0
+		b1 *= z.src_blend[x];
+		r2 *= z.dst_blend[x];
+		g2 *= z.dst_blend[x];
+		b2 *= z.dst_blend[x];
+		r1 += r2;
+		g1 += g2;
+		b1 += b2;
+
+		r1 >>= 3;
+		g1 >>= 3;
+		b1 >>= 3;
+		r1 = std::min(r1, static_cast<u16>(255));
+		g1 = std::min(g1, static_cast<u16>(255));
+		b1 = std::min(b1, static_cast<u16>(255));
```

Comment:
You don’t need `static_cast` – it’s more readable to use function style casts or explicit template arguments:

Like this:
```c++
		r1 = std::min(r1, u16(255));
		g1 = std::min(g1, u16(255));
		b1 = std::min(b1, u16(255));
```

Or this:
```c++
		r1 = std::min<u16>(r1, 255);
		g1 = std::min<u16>(g1, 255);
		b1 = std::min<u16>(b1, 255);
```

Reply:
_oh_